mitxgraders/baseclasses.py

```diff
@@ -187,6 +219,61 @@ def __call__(self, expect, student_input):
         self.format_messages(result)
         return result
 
+    def apply_attempt_based_credit(self, result, attempt_number):
+        """
+        Apply attempt-based credit maximums to grading.
+        Mutates result directly.
+        """
+        if attempt_number is None:
+            self.log("Attempt-based credit requested, but attempt number "
+                     "not passed through to grader")
+            return
+        self.log("Attempt number {}".format(attempt_number))
+
+        # How far past the point of decreasing credit are we?
+        if attempt_number < 1:  # Just in case edX has issues
+            attempt_number = 1
+        steps = attempt_number - self.config['decrease_credit_after']
+        if steps <= 0:
+            return
+
+        # Compute the credit to be awarded
+        min_cred = self.config['minimum_credit']
+        decrease_steps = self.config['decrease_credit_steps']
+        if steps >= decrease_steps:
+            credit = min_cred
+        else:
+            # Linear interpolation
+            credit = 1 + (min_cred - 1) * steps / decrease_steps
+        self.log("Maximum credit is {}".format(credit))
+
+        # Multiply all grades by credit, updating from 'ok'=True to 'partial' as needed
+        changed_result = False
+        if "input_list" in result:
+            for results_dict in result['input_list']:
+                if results_dict['grade_decimal'] > 0:
+                    grade = results_dict['grade_decimal'] * credit
+                    results_dict['grade_decimal'] = grade
+                    results_dict['ok'] = {0: False, 1: True}.get(grade, 'partial')
```

Comment:
ItemGrader has 

```python
@staticmethod
def grade_decimal_to_ok(grade):
    """Converts a grade decimal into an 'ok' value: True, False or 'partial'"""
    return {0: False, 1: True}.get(grade, 'partial')
```
you could move that into Abstract grader and use it here, if you want.

Reply:
Good point. Done.